urllib3/util/rfc6555.py

```diff
@@ -0,0 +1,241 @@
+                socks.append(sock)
+
+                # Using EVENT_WRITE to detect a connection.
+                selector.register(sock, EVENT_WRITE)
+
+            except (socket.error, OSError) as e:
+                err = e
+                if sock is not None:
+                    sock.close()
+
+            # This is where the selecting logic occurs. Once all sockets
+            # have been created and are added to the selector, they should
+            # be continually selected until one of them works.  If there are
+            # still sockets to add to the selector, then select for only 200
+            # ms before adding the next socket to the selector.
+            reattempt_select = True
+            while reattempt_select:
+                if i < dns_results_len - 1:
+                    # If this isn't the last socket to try, then only try
+                    # for 200 ms before adding the next preferred connection
+                    # to the selector. The 200 ms constant is the recommended
+                    # value by the RFC.
+                    select_time = 0.2
+                    reattempt_select = False  # Only want one round.
+                else:
+                    # Here we need to calculate how long we're willing to wait
+                    # as there are no sockets left to add to the selector.
+                    if timeout_time is not None:
+                        # If there are no more entries to add to the selector
+                        # we're going to keep trying until we either find
+                        # a socket without errors or we run out of time
+                        # to establish a connection.
+                        select_time = timeout_time - monotonic()
+
+                        # We do this for safety reasons, as negative timeout
+                        # may mean no timeout for certain selectors.
+                        if select_time < 0.0:
+                            break
+                    else:
+                        # Otherwise we're willing to block forever, but set a time
+                        # so that we can still filter out errored sockets.
+                        select_time = 1.0
+
+                # Make sure we don't have any errored sockets.
+                for sock in socks:
+                    errcode = sock.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
+                    if errcode and errcode not in _ASYNC_ERROR_NUMBERS:
+                        err = socket.error(errcode)
+                        socks.remove(sock)
+                        selector.unregister(sock)
+
+                # If we don't have any sockets, don't try selecting.
+                if not socks:
+                    # If there are no more sockets to look through, then
+                    # we should give up immediately with a non-timeout error.
+                    if i == dns_results_len - 1:
+                        err = socket.error(errno.ECONNREFUSED)
```

Comment:
Rather than use `ECONNREFUSED`, it would be better to use the last error we got from any of our sockets.